### PR TITLE
add oc adm ocp-certificates regenerate-top-level

### PR DESCRIPTION
**Pull request by deads2k**
replace https://github.com/openshift/oc/pull/1438

Adds a command that will cause operators to regenerate the self-signed signing certificates at the top of their chains.

[deads@fedora ~]$ oc adm ocp-certificates regenerate-signers -A secrets --all --dry-run 
secret/aggregator-client-signer regeneration set
secret/kube-apiserver-to-kubelet-signer regeneration set
secret/kube-control-plane-signer regeneration set
secret/loadbalancer-serving-signer regeneration set
secret/localhost-recovery-serving-signer regeneration set
secret/localhost-serving-signer regeneration set
secret/node-system-admin-signer regeneration set
secret/service-network-serving-signer regeneration set
secret/csr-signer-signer regeneration set
secret/ovn-ca regeneration set
secret/signer-ca regeneration set
Needs unit tests.

/assign @stlaz

### CODE REVIEW

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	for _, certificate := range certificates {
+		detail, err := toCertKeyPair(certificate)
+		if err != nil {
+			return nil, err
+		}
+		detail = addSecretLocation(detail, obj.Namespace, obj.Name)
+		return detail, nil
+	}
+	return nil, fmt.Errorf("didn't see that coming")
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/analyzer.go:35]: :eyes: 
does this intentionally skip all but the first cert?

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	"fmt"
+	"os"
+
+	"github.com/openshift/oc/pkg/cli/admin/ocpcertificates/certgraphapi"
+	"k8s.io/apimachinery/pkg/util/duration"
+)
+
+func toCertKeyPair(certificate *x509.Certificate) (*certgraphapi.CertKeyPair, error) {
+	ret := &certgraphapi.CertKeyPair{
+		Name: fmt.Sprintf("%v::%v", certificate.Subject.CommonName, certificate.SerialNumber),
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:16]: `Subject` has a `String()` method that might work better than just the common name

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	return ret, nil
+}
+
+func toCertKeyPairDetails(certificate *x509.Certificate) (certgraphapi.CertKeyPairDetails, error) {
+	count := 0
+	isClient := false
+	isServing := false
+	isSigner := (certificate.KeyUsage & x509.KeyUsageCertSign) != 0
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:37]: wouldn't `certificate.IsCA` work here?

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	ret.Spec.Details = details
+	if err != nil {
+		ret.Status.Errors = append(ret.Status.Errors, err.Error())
+	}
+
+	return ret, nil
+}
+
+func toCertKeyPairDetails(certificate *x509.Certificate) (certgraphapi.CertKeyPairDetails, error) {
+	count := 0
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:34]: `usageCount`

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	isClient := false
+	isServing := false
+	isSigner := (certificate.KeyUsage & x509.KeyUsageCertSign) != 0
+	if isSigner {
+		count++
+	}
+	for _, curr := range certificate.ExtKeyUsage {
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:41]: what happens with `ExtKeyUsageAny` ?

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		if curr == x509.ExtKeyUsageServerAuth {
+			isServing = true
+			count++
+		}
+	}
+	var typeError error
+	if count == 0 {
+		typeError = fmt.Errorf("you have a cert for nothing?")
+	}
+	if count > 1 {
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:55]: we do, in etcd peer certificates :neutral_face:

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+	if isServing {
+		ret.CertType = "ServingCertDetails"
+		ret.ServingCertDetails = toServingCertDetails(certificate)
+	}
+	if isSigner {
+		ret.CertType = "SignerCertDetails"
+		ret.SignerDetails = toSignerDetails(certificate)
+	}
+
+	if count > 1 {
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:73]: the same causes typeError above

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	return ret
+}
+
+func toSignerDetails(certificate *x509.Certificate) *certgraphapi.SignerCertDetails {
+	return &certgraphapi.SignerCertDetails{}
+}
+
+func toCertKeyMetadata(certificate *x509.Certificate) certgraphapi.CertKeyMetadata {
+	ret := certgraphapi.CertKeyMetadata{
+		CertIdentifier: certgraphapi.CertIdentifier{
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:105]: a proper identifier would consist of all of:
Subject
Issuer
SerialNumber

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	switch publicKey := certificate.PublicKey.(type) {
+	case *ecdsa.PublicKey:
+		ret.PublicKeyBitSize = fmt.Sprintf("%d bit, %v curve", publicKey.Params().BitSize, publicKey.Params().Name)
+	case *rsa.PublicKey:
+		ret.PublicKeyBitSize = fmt.Sprintf("%d bit", publicKey.Size()*8)
+	default:
+		fmt.Fprintf(os.Stderr, "%T\n", publicKey)
+	}
+
+	signerHumanName := certificate.Issuer.CommonName
```
Comment (stlaz) [inline pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:123]: use the full `Issuer.String()`

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 3 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:37]: >wouldn't certificate.IsCA work here?

I don't know.  Test before a followup to switch.  Also test in cluster-debug-tools please.

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/analyzer.go (hunk 1 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/analyzer.go:35]: > does this intentionally skip all but the first cert?

My memory is that *our* certificates always list the thing we're interested in first and the tool is for our platform, not general purpose.

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 5 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:41]: > what happens with ExtKeyUsageAny ?

We can update to make it do something in a followup. This is for inspecting our platform certificates, so the use-cases are constrained.

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 6 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:55]: Yeah, see marker in https://github.com/openshift/api/blob/master/tls/docs/etcd%20Certificates/cert-flow.json#L510 .

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 7 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:73]: > the same causes typeError above

errors here are reported as marker, not as fatal "do not progress".  Open to refactors after cluster-debug-tools aligns.

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 8 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:105]: Subject versus CommonName?  Refactor welcome after vendoring into cluster-debug-tools and seeing outcomes.

File: pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go (hunk 9 above)
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go:123]: If memory serves, that did bad things in the generated doc, but try it in cluster-debug-tools against a cluster and we can tweak here if appropriate.

File: pkg/cli/admin/ocpcertificates/regeneratesigners/regenerate.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	if issuerInfo == nil {
+		// not for this command.
+		return nil
+	}
+
+	if issuerInfo.CommonName != keyPairInfo.Spec.CertMetadata.CertIdentifier.CommonName {
+		// not for this command, we only want self-signed signers.
+		//fmt.Printf("#### SKIPPING ns/%v secret/%v issuer=%v\n", secret.Namespace, secret.Name, keyPairInfo.Spec.CertMetadata.CertIdentifier.Issuer)
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cli/admin/ocpcertificates/regeneratesigners/regenerate.go]: @stlaz pushed an example of a unit test.  I'm not seeing the value in the separation.

**Comment by stlaz** [comment]
chose a little bit of a different approach for unit testing - https://github.com/stlaz/oc/commit/603e2451d1612693eaa7733f5ed8bed4c8af9e12

File: pkg/cli/admin/ocpcertificates/regeneratesigners/regenerate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/cli/admin/ocpcertificates/regeneratesigners/regenerate.go]: Having the logic in a separate unit for testing just that unit would make sense, you'd test the client behavior in a different unit test.

Does not matter, it's just a nit. The unit test for the inner-most function are to be found at https://github.com/stlaz/oc/commit/603e2451d1612693eaa7733f5ed8bed4c8af9e12

**Comment by deads2k** [comment]
> chose a little bit of a different approach for unit testing - [stlaz@603e245](https://github.com/stlaz/oc/commit/603e2451d1612693eaa7733f5ed8bed4c8af9e12)

Added.  Thanks.

**Comment by stlaz** [comment]
/lgtm
good enough for now

**Pull request merged**